Structures moving out of the domain

nayaksagar opened the pull request:
<!--
This template should be included in all pull requests. Items in the list should
either be completed by the original author or explicitly dismissed by one of the
IBAMR principal developers.

IBAMR is a community effort and it wouldn't exist without people contributing
code. Thanks in advance for helping to make IBAMR better!
-->

### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

## Description
Examples in the 'constraintIBMethod' throw error when structures try to move out of the computational domain through an outlet boundary. In this patch 'inactiveLagrangianStructures' functionality of 'LDataManager'  is used to deactivate the structure as it approaches the outlet boundary. Changes are made in the 'fallingSphere' problem so as to let the sphere out of the domain. Bottom boundary of the computational domain is given 'outlet boundary condition' to realize this scenario.

![sphere_with_lag](https://github.com/IBAMR/IBAMR/assets/88955027/fd94d0f4-bcd9-4145-a7ff-dc726350c42d)

nayaksagar commented:
Examples in the 'constraintIBMethod' throw error when structures try to move out of the computational domain through an outlet boundary. In this patch 'inactiveLagrangianStructures' functionality of 'LDataManager' is used to deactivate the structure as it approaches the outlet boundary. Changes are made in the 'fallingSphere' problem so as to let the sphere out of the domain. Bottom boundary of the computational domain is given 'outlet boundary condition' to realize this scenario.

amneetb commented on examples/ConstraintIB/falling_sphere/example.cpp:
```diff
@@ -282,7 +282,6 @@ main(int argc, char* argv[])
             for(int structure_idx = 0; structure_idx < structure_COM.size(); structure_idx++){
                 if( (structure_COM[structure_idx][1]-radius) < x_lo[1]){ // changes based on where "open boundaries" are present
```
We need to check for the open boundary condition here. This can be checked either as a flag from input file or checking the coefficients of the normal component of velocity.

amneetb commented on examples/ConstraintIB/falling_sphere/input3d:
```diff
@@ -66,14 +66,14 @@ VelocityBcCoefs_1 {
    bcoef_function_2 = "1.0"
-   bcoef_function_3 = "1.0"
+   bcoef_function_3 = "0.0"
    bcoef_function_4 = "0.0"
```
Suggest adding a flag saying if we want to do an open or closed boundary here:

```
...
OPEN_BOUNDARY = TRUE
...
UCoef_db
{
 
open_boundary_3 = OPEN_BOUNDARY

...

acoef_function_3 = "open_boundary? 0.0 : 1.0"

bcoef_function_3 = ""open_boundary? 1.0 : 0.0""
 
}

```

amneetb commented:
> Thanks! I have a suggestion to clean things up.
> 
> @amneetb Would it make sense to make this a feature of `ConstraintIB` in general instead of this example? Should we modify the input file?

No I think this case dependent and should be done through main.cpp. I provided a suggestion for the input file to handle both the closed and open boundary cases.

nayaksagar commented:
Hello Dr. Wells, Dr. Amneet,
Sorry for the delayed reply. I somehow missed this conversation. Thanks for
the suggestions. I am working on them and will update them soon.

Thanks

On Thu, Jan 4, 2024 at 10:44 PM David Wells ***@***.***>
wrote:

> @nayaksagar <https://github.com/nayaksagar> Can you make the suggested
> changes? It would be great to get this in for the next release (which is
> due out by the end of the month).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/IBAMR/IBAMR/pull/1658#issuecomment-1877473820>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVGVRE2G2ZO4NTY4S5NHAZTYM3POPAVCNFSM6AAAAABAMMSTUCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNZXGQ3TGOBSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
*Regards*
Sagar G. Nayak
Research Scholar,
University of Queensland-IIT Delhi Academy of Research (UQIDAR),
Indian Institute of Technology Delhi,
Hauz Khas, New Delhi-110016, India.
ORCID: https://orcid.org/0000-0001-6129-571X

nayaksagar commented:
@drwells and @amneetb. Please see the changes. I took the suggestion of checking the coefficients of normal component of velocity to see whether there are any open boundaries. Code checks where there is an open boundary and allows the structure to go out from there. 
I checked by changing the gravity direction and making the corresponding wall open.
gx
![gx](https://github.com/IBAMR/IBAMR/assets/88955027/30057a92-fb6b-483d-ae08-aed0abd2806a)
gy
![gy](https://github.com/IBAMR/IBAMR/assets/88955027/a77fd0d7-0fc0-418b-9ad5-6445c7175693)
gz
![gz](https://github.com/IBAMR/IBAMR/assets/88955027/9a8326ce-f9b3-4ba9-8f4b-fb087ea5c3c7)

Also changed the number of ghost points in ForceProjector.cpp (earlier it was given 0)

I would like to know your comments. Thanks

boyceg commented:
@drwells is there anything that still needs to be done on this PR?

boyceg commented:
I guess these are approved by @drwells but not yet from @amneetb -- @amneetb can you take a look?

amneetb commented:
Sorry for the late review:

 Can you replace "hi" with a more formal output message?

https://github.com/IBAMR/IBAMR/pull/1658/commits/f87bd684b61b97d5bec9e22585f98ee31d683569#diff-b9951955fcaba91558694ef53c561c01fe732aaa62bc665fc9ab1ed4c58d04f6R302

amneetb commented:
This example corresponds to the problem setup of our JCP 2013 paper. I would suggest not to (unless there is a particular reason for doing this change) change the initial position or center of mass of the sphere, or direction of gravity so that there is one-to-one correspondence.

nayaksagar commented:
Thank you Dr Amneet. 
I think there is a slight confusion (this might have also caused because of me being new to Git). I have already made the changes that you are suggesting.
I have removed the unnecessary comments in the commit https://github.com/IBAMR/IBAMR/pull/1658/commits/0ecdf0bc6bf1f9f0acfadff1ae45440f9dfcfc78,
have changed the gravity back to original direction in https://github.com/IBAMR/IBAMR/pull/1658/commits/2afe6beb3a5f637d8a23de3f4ffdd06615ecd139,
and changed the sphere coordinates back to original position in https://github.com/IBAMR/IBAMR/pull/1658/commits/ad0489fbefd221685996a1341c7f447c23fd9522

Sorry for not including them all in one commit. Please let me know if you cannot find these commits.
Thank you

nayaksagar commented:
@amneetb: Thank you for explaining Dr Amneet. I have now squashed all the commits and pushed it. Please see if it is clear now.
Thanks

drwells commented:
@amneetb is this ready?

nayaksagar commented:
Hello Dr Amneet, can you please check if this PR needs any changes. I can work on it

boyceg commented:
@amneetb, what do you think, is this one good to merge?
